framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -5668,4 +5668,32 @@ public boolean isImmutable(TypeMirror type) {
         }
         return ImmutableTypes.isImmutable(TypeAnnotationUtils.unannotatedType(type).toString());
     }
+
+    /**
+     * Returns the qualifiers in {@code annos} that are below the qualifier upper bound of {@code
+     * type}. If a qualifier in {@code annos} is above the bound, then the bound is added to the
+     * result instead.
+     *
+     * @param type java type that specifies the qualifier upper bound
+     * @param annos annotations to add to the {@link AnnotatedTypeMirror} of type
+     * @return the modified {@code annos} by applying the rules above
+     */
+    public Set<AnnotationMirror> getAnnoOrTypeBound(
+            TypeMirror type, Set<? extends AnnotationMirror> annos) {
+        Set<AnnotationMirror> boundAnnos = getTypeDeclarationBounds(type);
+        Set<AnnotationMirror> results = AnnotationUtils.createAnnotationSet();
+
+        for (AnnotationMirror anno : annos) {
```

Comment:
Shouldn't this iterate over `boundAnnos` - like the original code in `addAnnoOrBound` did?
Otherwise, `annos` could be fewer annotations than `boundAnnos` and miss some?
Or at least specify what the desired behavior for the method is.

Reply:
Added documentation about the desired behavior for the method.